FFMDigitalInput$EventWatcher [ERROR]

Test environment: Java 25, pi4J 4.0, Raspberry Pi 4B rev 1.2, Debian rel. 13


Running a simple test configuring an IO as input with a listener. Getting this error when calling pi4j.shutdown();
--> com.pi4j.plugin.ffm.providers.gpio.FFMDigitalInput$EventWatcher [ERROR] [/dev/gpiochip0-event-detection-pin-21] /dev/gpiochip0-event-detection-pin-21 - **Internal error during polling on BCM 21**. Last polling data: PollingData{fd=15, events=3, revents=32}

source code here:
[TestAppGPIOdebug.java](https://github.com/user-attachments/files/26010927/TestAppGPIOdebug.java)

dependencies:
[pom.xml](https://github.com/user-attachments/files/26010941/pom.xml)

Is attaching a listener to an input create a polling mechanism as the error message indicate, or is it interrupt based?
thanks for looking at this.

Max

--- a/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/gpio/FFMDigitalInput.java
+++ b/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/gpio/FFMDigitalInput.java
@@ -150,9 +150,6 @@ public class FFMDigitalInput extends DigitalInputBase implements DigitalInput {
         super.shutdownInternal(context);
         logger.info("{}-{} - closing GPIO BCM.", deviceName, bcm);
         try {
-            if (chipFileDescriptor > 0) {
-                file.close(chipFileDescriptor);
-            }
             logger.trace("{}-{} - Stopping event watchers", deviceName, bcm);
             for (EventWatcher watcher : watchers) {
                 watcher.stopWatching();
@@ -168,9 +165,14 @@ public class FFMDigitalInput extends DigitalInputBase implements DigitalInput {
         } catch (Exception e) {
             this.closed = true;
             throw new ShutdownException(e);
+        } finally {
+            if (chipFileDescriptor > 0) {
+                logger.trace("{}-{} - closing GPIO file descriptor '{}'.", deviceName, bcm, chipFileDescriptor);
+                file.close(chipFileDescriptor);
+            }
         }
         this.closed = true;
-        logger.info("{}-{} - GPIO BCM is closed. Recreate the pin object to reuse.", deviceName, bcm);
+        logger.info("{}-{} - GPIO BCM is closed.", deviceName, bcm);
         return this;
     }
 

